[BUG] Micronaut 3.0: Can't use @PostConstruct in ConfigurationProperties beans

### Expected Behavior

_No response_

### Actual Behaviour

Try to start sample appication and you will see this exception:
```
19:53:31.644 ERROR [main] i.m.runtime.Micronaut [Micronaut.java:326] - Error starting Micronaut server: Error instantiating bean of type  [com.micronaut.bug.service.entity.service.EntityService]

Path Taken: new EntityService(EntityProperties entityProperties) --> new EntityService([EntityProperties entityProperties])
io.micronaut.context.exceptions.BeanInstantiationException: Error instantiating bean of type  [com.micronaut.bug.service.entity.service.EntityService]

Path Taken: new EntityService(EntityProperties entityProperties) --> new EntityService([EntityProperties entityProperties])
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2265)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:3193)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:3179)
	at io.micronaut.context.DefaultBeanContext.getBeanForDefinition(DefaultBeanContext.java:2723)
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2685)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1555)
	at io.micronaut.context.AbstractInitializableBeanDefinition.resolveBean(AbstractInitializableBeanDefinition.java:1561)
	at io.micronaut.context.AbstractInitializableBeanDefinition.getBeanForConstructorArgument(AbstractInitializableBeanDefinition.java:1044)
	at com.micronaut.bug.service.entity.service.$EntityService$Definition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2238)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:3193)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:3179)
	at io.micronaut.context.DefaultBeanContext.getBeanForDefinition(DefaultBeanContext.java:2723)
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2685)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:803)
	at io.micronaut.context.DefaultBeanContext$BeanExecutionHandle.getTarget(DefaultBeanContext.java:3892)
	at io.micronaut.context.DefaultBeanContext$BeanExecutionHandle.invoke(DefaultBeanContext.java:3913)
	at io.micronaut.aop.chain.AdapterIntroduction.intercept(AdapterIntroduction.java:83)
	at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:137)
	at com.micronaut.bug.service.entity.service.EntityService$ApplicationEventListener$init1$Intercepted.onApplicationEvent(Unknown Source)
	at io.micronaut.context.event.ApplicationEventPublisherFactory.notifyEventListeners(ApplicationEventPublisherFactory.java:262)
	at io.micronaut.context.event.ApplicationEventPublisherFactory.access$200(ApplicationEventPublisherFactory.java:60)
	at io.micronaut.context.event.ApplicationEventPublisherFactory$2.publishEvent(ApplicationEventPublisherFactory.java:229)
	at io.micronaut.context.DefaultBeanContext.publishEvent(DefaultBeanContext.java:1650)
	at io.micronaut.http.server.netty.NettyHttpServer.fireStartupEvents(NettyHttpServer.java:560)
	at io.micronaut.http.server.netty.NettyHttpServer.start(NettyHttpServer.java:403)
	at io.micronaut.http.server.netty.NettyHttpServer.start(NettyHttpServer.java:155)
	at io.micronaut.runtime.Micronaut.lambda$start$2(Micronaut.java:77)
	at java.base/java.util.Optional.ifPresent(Optional.java:183)
	at io.micronaut.runtime.Micronaut.start(Micronaut.java:75)
	at com.micronaut.bug.Application.main(Application.java:12)
Caused by: java.lang.NullPointerException: null
	at io.micronaut.context.AbstractInitializableBeanDefinition.containsValueForMethodArgument(AbstractInitializableBeanDefinition.java:921)
	at com.micronaut.bug.service.entity.config.$EntityProperties$Definition.injectBean(Unknown Source)
	at com.micronaut.bug.service.entity.config.$EntityProperties$Definition.build(Unknown Source)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2238)
	... 30 common frames omitted
```

### Steps To Reproduce

Problem in class `com.micronaut.bug.service.entity.config.EntityProperties` when I use `@PostConstruct` annotation:
```
@Slf4j
@Getter
@Setter
@ConfigurationProperties("app.entity")
public class EntityProperties {

    private String prop;

    /**
     * !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
     * That is PROBLEM block. If you comment this - all works fine
     */
    @PostConstruct
    public void init() {
        log.info("prop = {}", prop);
    }
}
```

You can see full sample application to reproduce this bug.

PS https://github.com/micronaut-projects/micronaut-data/issues/973 - this problem isn't solved yet too :-(

### Environment Information

Win10
JDK 11
Gradle 7.2


### Example Application

https://github.com/altro3/micronaut3-bug

### Version

3.0.0

All works fine with Micronaut 2.5

--- a/inject-java/src/test/groovy/io/micronaut/inject/configproperties/ConfigPropertiesParseSpec.groovy
+++ b/inject-java/src/test/groovy/io/micronaut/inject/configproperties/ConfigPropertiesParseSpec.groovy
@@ -1,6 +1,7 @@
 package io.micronaut.inject.configproperties
 
 import io.micronaut.context.ApplicationContext
+import io.micronaut.context.BeanContext
 import io.micronaut.context.annotation.ConfigurationReader
 import io.micronaut.context.annotation.Property
 import io.micronaut.core.convert.format.ReadableBytes
@@ -870,5 +871,40 @@ class Parent {
         beanDefinition != null
         beanDefinition.getAnnotationMetadata().stringValue(ConfigurationReader.class, "prefix").get() == "parent.children[*]"
         beanDefinition.getRequiredMethod("getPropA").getAnnotationMetadata().getAnnotationValuesByType(Property.class).get(0).stringValue("name").get() == "parent.children[*].prop-a"
+    }
+
+    void "test config props with post construct first in file"() {
+        given:
+        BeanContext context = buildContext("test.EntityProperties", """
+package test;
+
+import io.micronaut.context.annotation.ConfigurationProperties;
+import jakarta.annotation.PostConstruct;
+
+@ConfigurationProperties("app.entity")
+public class EntityProperties {
+
+    private String prop;
+    
+    @PostConstruct
+    public void init() {
+        System.out.println("prop = " + prop);
+    }
+    
+    public String getProp() {
+        return prop;
+    }
+    
+    public void setProp(String prop) {
+        this.prop = prop;
+    }
+}
+""")
+
+        when:
+        context.getBean(context.classLoader.loadClass("test.EntityProperties"))
+
+        then:
+        noExceptionThrown()
     }
 }

--- a/inject/src/main/java/io/micronaut/inject/writer/BeanDefinitionWriter.java
+++ b/inject/src/main/java/io/micronaut/inject/writer/BeanDefinitionWriter.java
@@ -382,6 +382,7 @@ public class BeanDefinitionWriter extends AbstractClassFileWriter implements Bea
     private final List<MethodVisitData> methodInjectionPoints = new ArrayList<>(2);
     private final List<MethodVisitData> postConstructMethodVisits = new ArrayList<>(2);
     private final List<MethodVisitData> preDestroyMethodVisits = new ArrayList<>(2);
+    private final List<MethodVisitData> allMethodVisits = new ArrayList<>(2);
     private final Map<String, Boolean> isLifeCycleCache = new HashMap<>(2);
     private ExecutableMethodsDefinitionWriter executableMethodsDefinitionWriter;
 
@@ -869,42 +870,21 @@ public class BeanDefinitionWriter extends AbstractClassFileWriter implements Bea
         classWriter.visitField(ACC_PRIVATE | ACC_FINAL | ACC_STATIC, FIELD_CONSTRUCTOR,
                 Type.getType(AbstractInitializableBeanDefinition.MethodOrFieldReference.class).getDescriptor(), null, null);
 
-        int methodsLength = methodInjectionPoints.size() + postConstructMethodVisits.size() + preDestroyMethodVisits.size();
+        int methodsLength = allMethodVisits.size();
         if (!superBeanDefinition && methodsLength > 0) {
             Type methodsFieldType = Type.getType(AbstractInitializableBeanDefinition.MethodReference[].class);
             classWriter.visitField(ACC_PRIVATE | ACC_FINAL | ACC_STATIC, FIELD_INJECTION_METHODS, methodsFieldType.getDescriptor(), null, null);
             pushNewArray(staticInit, AbstractInitializableBeanDefinition.MethodReference.class, methodsLength);
             int i = 0;
-            for (MethodVisitData methodVisitData : methodInjectionPoints) {
+            for (MethodVisitData methodVisitData : allMethodVisits) {
                 pushStoreInArray(staticInit, i++, methodsLength, () ->
                         pushNewMethodReference(
                                 staticInit,
                                 JavaModelUtils.getTypeReference(methodVisitData.beanType),
                                 methodVisitData.methodElement,
                                 methodVisitData.requiresReflection,
-                                false, false
-                        )
-                );
-            }
-            for (MethodVisitData methodVisitData : postConstructMethodVisits) {
-                pushStoreInArray(staticInit, i++, methodsLength, () ->
-                        pushNewMethodReference(
-                                staticInit,
-                                JavaModelUtils.getTypeReference(methodVisitData.beanType),
-                                methodVisitData.methodElement,
-                                methodVisitData.requiresReflection,
-                                true, false
-                        )
-                );
-            }
-            for (MethodVisitData methodVisitData : preDestroyMethodVisits) {
-                pushStoreInArray(staticInit, i++, methodsLength, () ->
-                        pushNewMethodReference(
-                                staticInit,
-                                JavaModelUtils.getTypeReference(methodVisitData.beanType),
-                                methodVisitData.methodElement,
-                                methodVisitData.requiresReflection,
-                                false, true
+                                methodVisitData.isPostConstruct(),
+                                methodVisitData.isPreDestroy()
                         )
                 );
             }
@@ -1171,6 +1151,7 @@ public class BeanDefinitionWriter extends AbstractClassFileWriter implements Bea
         );
 
         methodInjectionPoints.add(methodVisitData);
+        allMethodVisits.add(methodVisitData);
 
         if (!requiresReflection) {
             resolveBeanOrValueForSetter(declaringType, methodElement, isOptional);
@@ -1187,8 +1168,9 @@ public class BeanDefinitionWriter extends AbstractClassFileWriter implements Bea
         visitPostConstructMethodDefinition(false);
         // for "super bean definitions" we just delegate to super
         if (!superBeanDefinition || isInterceptedLifeCycleByType(this.annotationMetadata, "POST_CONSTRUCT")) {
-            MethodVisitData methodVisitData = new MethodVisitData(declaringType, methodElement, requiresReflection);
+            MethodVisitData methodVisitData = new MethodVisitData(declaringType, methodElement, requiresReflection, true, false);
             postConstructMethodVisits.add(methodVisitData);
+            allMethodVisits.add(methodVisitData);
             visitMethodInjectionPointInternal(methodVisitData, postConstructMethodVisitor, postConstructInstanceLocalVarIndex);
         }
     }
@@ -1202,8 +1184,9 @@ public class BeanDefinitionWriter extends AbstractClassFileWriter implements Bea
         if (!superBeanDefinition || isInterceptedLifeCycleByType(this.annotationMetadata, "PRE_DESTROY")) {
             visitPreDestroyMethodDefinition(false);
 
-            MethodVisitData methodVisitData = new MethodVisitData(declaringType, methodElement, requiresReflection);
+            MethodVisitData methodVisitData = new MethodVisitData(declaringType, methodElement, requiresReflection, false, true);
             preDestroyMethodVisits.add(methodVisitData);
+            allMethodVisits.add(methodVisitData);
             visitMethodInjectionPointInternal(methodVisitData, preDestroyMethodVisitor, preDestroyInstanceLocalVarIndex);
         }
     }
@@ -1216,6 +1199,7 @@ public class BeanDefinitionWriter extends AbstractClassFileWriter implements Bea
 
         MethodVisitData methodVisitData = new MethodVisitData(declaringType, methodElement, requiresReflection);
         methodInjectionPoints.add(methodVisitData);
+        allMethodVisits.add(methodVisitData);
         visitMethodInjectionPointInternal(methodVisitData, injectMethodVisitor, injectInstanceLocalVarIndex);
     }
 
@@ -3072,7 +3056,7 @@ public class BeanDefinitionWriter extends AbstractClassFileWriter implements Bea
                 protectedConstructor.getStatic(getTypeReferenceForName(getBeanDefinitionReferenceClassName()), AbstractAnnotationMetadataWriter.FIELD_ANNOTATION_METADATA, Type.getType(AnnotationMetadata.class));
             }
             // 4: `AbstractBeanDefinition2.MethodReference[].class` methodInjection
-            if (methodInjectionPoints.isEmpty() && preDestroyMethodVisits.isEmpty() && postConstructMethodVisits.isEmpty()) {
+            if (allMethodVisits.isEmpty()) {
                 protectedConstructor.push((String) null);
             } else {
                 protectedConstructor.getStatic(beanDefinitionType, FIELD_INJECTION_METHODS, Type.getType(AbstractInitializableBeanDefinition.MethodReference[].class));
@@ -3510,6 +3494,8 @@ public class BeanDefinitionWriter extends AbstractClassFileWriter implements Bea
         private final TypedElement beanType;
         private final boolean requiresReflection;
         private final MethodElement methodElement;
+        private final boolean postConstruct;
+        private final boolean preDestroy;
 
         /**
          * Default constructor.
@@ -3525,6 +3511,21 @@ public class BeanDefinitionWriter extends AbstractClassFileWriter implements Bea
             this.beanType = beanType;
             this.requiresReflection = requiresReflection;
             this.methodElement = methodElement;
+            this.postConstruct = false;
+            this.preDestroy = false;
+        }
+
+        MethodVisitData(
+                TypedElement beanType,
+                MethodElement methodElement,
+                boolean requiresReflection,
+                boolean postConstruct,
+                boolean preDestroy) {
+            this.beanType = beanType;
+            this.requiresReflection = requiresReflection;
+            this.methodElement = methodElement;
+            this.postConstruct = postConstruct;
+            this.preDestroy = preDestroy;
         }
 
         /**
@@ -3547,7 +3548,16 @@ public class BeanDefinitionWriter extends AbstractClassFileWriter implements Bea
         public boolean isRequiresReflection() {
             return requiresReflection;
         }
+
+        public boolean isPostConstruct() {
+            return postConstruct;
+        }
+
+        public boolean isPreDestroy() {
+            return preDestroy;
+        }
     }
+
 
     private class FactoryMethodDef {
         private final Type factoryType;

